mixer/cmd/root.go

```diff
@@ -54,11 +58,33 @@ var initCmd = &cobra.Command{
 	},
 }
 
+var parseErrorPrefixes = []string{
+	"unknown command \"",
+	"unknown flag:",
+	"unknown shorthand flag:",
+	"invalid argument \"",
+	"flag needs an argument:",
+	"bad flag syntax:",
+}
+
 // Execute adds all child commands to the root command and sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
 func Execute() {
-	if err := RootCmd.Execute(); err != nil {
-		fmt.Fprintf(os.Stderr, "Mixer Error: %s\n", err)
+	cmd, err := RootCmd.ExecuteC()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
+
+		// Automatically show usage information if it is a flag parsing error.  Flag parsing
+		// doesn't return errors that we can use types or interfaces to match, so do an
+		// approximation.
+		s := err.Error()
+		for _, p := range parseErrorPrefixes {
+			if strings.HasPrefix(s, p) {
```

Comment:
Parsing and matching error strings seems so wrong :(

Reply:
I've added an comment above explaining why we can't do the nice way. If pflags returned errors from a specific type, we could switch on type here and do a cleaner job, but they don't. 

cobra controls the parsing of the flags and when it happens, so we can't really do much from the outside. One alternative would be looking at replacing the UsageFunc for the commands and print from there or store the information. Would have to dig deeper in how cobra uses it and I didn't think was worthwhile.

Given that "failing" here is not critical (i.e. we either miss showing a Usage message or show one extra Usage message), I guess we should be fine.

That said, I'm also happy to take this out and just use the approach from the first patch. I just don't want SPAM on every error :-)

Reply:
I really like the goal of this, but I agree with @tmarcu that matching on error string content just feels too hacky and fragile. You're absolutely right about switching on error types -- I think having the ability to discern parsing errors from others would be a useful enough feature that I think I might actually go submit an Issue on their pflag library asking for it.

I think I'd rather lose the automatic usage output altogether until a more reliable solution is possible.

Reply:
Oooo, I found it! There's a method called [SetFlagErrorFunc](https://godoc.org/github.com/spf13/cobra#Command.SetFlagErrorFunc). By default, its companion `FlagErrorFunc` is a noop, returning a function that just returns the original error. But you can override this to do whatever you want... like print the usage!

In `root.go`'s `init()`, insert this:

```go
	RootCmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
		cmd.Print(cmd.UsageString())
		return err
	})
```

Now usage is supressed for all errors thanks to `SilenceUsage`, but if a _flag_ error occurs, it prints the ussage and then passes it back up. Produces output like:

```
root@machine /home/clr/mix # mixer add-rpms -t
Mixer 3.2.1
Usage:
  mixer add-rpms [flags]

Flags:
  -c, --config string   Builder config to use
  -h, --help            help for add-rpms
Mixer Error: unknown shorthand flag: 't' in -t
```

Reply:
Great! I guess I should’ve keep digging. I’ll update the PR and test tomorrow.

Reply:
Fair warning: this doesn't trigger when other types of "usage" errors occur -- in other words, usages caught by Cobra itself, rather than pflag.

For example, if you type `mixer foobar` -- a _command_ usage error, not a _flag_ usage error -- you still get the error message with no usage info attached. Similarly, _argument_ errors are not caught with this, so if we took advantage of Cobra's "this command must have between 2 and 4 arguments in addition to any flags" type checks, that wouldn't spit usage either. (Potentially that could be caught with use of the [ValidateArgs](https://godoc.org/github.com/spf13/cobra#Command.ValidateArgs) function?)

I'm not too worried about these other cases, and the above flag error function covers a lot. And as @cmarcelo already pointed out, it's not like the malformed usage isn't being caught, it's just not resulting in a usage message.